## Day 1

In [2]:
import pandas as pd
from scipy.io import arff

# Load the arff file
data = arff.loadarff('../data/raw/pc1.arff')

# Convert to a pandas DataFrame
df = pd.DataFrame(data[0])

print(df.head())

     loc   v(g)  ev(g)  iv(G)       N         V     L      D       I  \
0    1.1    1.4    1.4    1.4     1.3      1.30  1.30   1.30    1.30   
1    1.0    1.0    1.0    1.0     1.0      1.00  1.00   1.00    1.00   
2   91.0    9.0    3.0    2.0   318.0   2089.21  0.04  27.68   75.47   
3  109.0   21.0    5.0   18.0   381.0   2547.56  0.04  28.37   89.79   
4  505.0  106.0   41.0   82.0  2339.0  20696.93  0.01  75.93  272.58   

            E  ...  lOCode  lOComment  locCodeAndComment  lOBlank  uniq_Op  \
0        1.30  ...     2.0        2.0                2.0      2.0      1.2   
1        1.00  ...     1.0        1.0                1.0      1.0      1.0   
2    57833.24  ...    80.0       44.0               11.0     31.0     29.0   
3    72282.68  ...    97.0       41.0               12.0     24.0     28.0   
4  1571506.88  ...   457.0       71.0               48.0     49.0     64.0   

   uniq_Opnd  total_Op  total_Opnd  branchCount   defects  
0        1.2       1.2         1.2    

In [3]:
print("Shape:")
print(df.shape)

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isnull().sum())

Shape:
(1109, 22)

Column names:
['loc', 'v(g)', 'ev(g)', 'iv(G)', 'N', 'V', 'L', 'D', 'I', 'E', 'B', 'T', 'lOCode', 'lOComment', 'locCodeAndComment', 'lOBlank', 'uniq_Op', 'uniq_Opnd', 'total_Op', 'total_Opnd', 'branchCount', 'defects']

Data types:
loc                  float64
v(g)                 float64
ev(g)                float64
iv(G)                float64
N                    float64
V                    float64
L                    float64
D                    float64
I                    float64
E                    float64
B                    float64
T                    float64
lOCode               float64
lOComment            float64
locCodeAndComment    float64
lOBlank              float64
uniq_Op              float64
uniq_Opnd            float64
total_Op             float64
total_Opnd           float64
branchCount          float64
defects               object
dtype: object

Missing values:
loc                  0
v(g)                 0
ev(g)                0
iv(G)      

In [4]:
df["defects"].value_counts()

defects
b'false'    1032
b'true'       77
Name: count, dtype: int64

In [5]:
df["defects"].value_counts(normalize=True) * 100

defects
b'false'    93.056808
b'true'      6.943192
Name: proportion, dtype: float64

In [6]:
df.describe()

,loc,v(g),ev(g),iv(G),N,V,L,D,I,E,...,T,lOCode,lOComment,locCodeAndComment,lOBlank,uniq_Op,uniq_Opnd,total_Op,total_Opnd,branchCount
count,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1.109000e+03,...,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000
mean,23.376105,5.510730,2.766817,3.321371,117.393417,699.711217,0.129414,15.396267,32.904472,2.882288e+04,...,1601.273030,22.434626,4.696123,0.944995,6.745717,13.307665,20.892876,66.493417,50.901894,9.576555
std,35.284017,8.958976,5.567704,6.401998,197.336888,1509.545684,0.146952,16.337508,35.387685,1.706436e+05,...,9480.199989,33.574645,10.518447,3.345297,12.301210,8.182224,29.051434,111.703236,86.308689,16.540676
min,0.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,...,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,1.000000,0.000000,1.000000
25%,7.000000,1.000000,1.000000,1.000000,25.000000,97.670000,0.050000,5.710000,14.040000,5.581300e+02,...,31.010000,7.000000,0.000000,0.000000,1.000000,8.000000,6.000000,15.000000,10.000000,1.000000
50%,13.000000,3.000000,1.000000,2.000000,58.000000,275.100000,0.080000,11.610000,23.570000,3.189170e+03,...,177.180000,13.000000,0.000000,0.000000,2.000000,12.000000,12.000000,33.000000,24.000000,5.000000
75%,26.000000,6.000000,3.000000,3.000000,126.000000,674.040000,0.160000,20.500000,41.150000,1.238156e+04,...,687.860000,24.000000,5.000000,1.000000,8.000000,17.000000,25.000000,72.000000,56.000000,11.000000
max,602.000000,136.000000,123.000000,123.000000,2785.000000,25942.690000,2.000000,270.660000,598.330000,4.279633e+06,...,237757.400000,600.000000,159.000000,48.000000,225.000000,99.000000,538.000000,1641.000000,1144.000000,236.000000


## Initial EDA Findings
* The class distribution of the data set is imbalanced.
* The PC1 dataset consists of 1109 rows and 22 columns. Rows contain information on software modules and columns contain attributes of teh respective software modules.
* There are 22 attribute columns out of which 21 columns are used as explanatory variates (input of the model) and 1 is response variate (output of the model).
* There are no missing values found in the dataset.
* There are total of 1032 non-defective modules and 77 defective modules out of 1109 software modules.
* Several explanatory variables are rightly skewed because their means are larger than their medians.

## Day 2

In [7]:
X = df.drop(columns=["defects"])
Y = df["defects"]

print(X.shape)
print(Y.shape)

(1109, 21)
(1109,)


In [8]:
from sklearn.model_selection import train_test_split

In [9]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42, stratify=Y)

In [10]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(887, 21)
(222, 21)
(887,)
(222,)


In [11]:
Y.value_counts(normalize=True)
Y_train.value_counts(normalize=True)
Y_test.value_counts(normalize=True)

defects
b'false'    0.932432
b'true'     0.067568
Name: proportion, dtype: float64

## Day 3

In [12]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline

In [13]:
Y_train = Y_train.astype(str)
Y_test = Y_test.astype(str)

dummy = DummyClassifier(strategy="most_frequent")

In [14]:
dummy.fit(X_train, Y_train)
dummy_prediction = dummy.predict(X_test)

print(dummy_prediction[:20])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']


In [15]:
dummyAccuracy = accuracy_score(Y_test, dummy_prediction)
dummyConfusionMatrix = confusion_matrix(Y_test, dummy_prediction)

print("Accuracy:", dummyAccuracy)
print("Confusion Matrix:")
print(dummyConfusionMatrix)

Accuracy: 0.9324324324324325
Confusion Matrix:
[[207   0]
 [ 15   0]]


In [16]:
precision = precision_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)
recall = recall_score(Y_test, dummy_prediction, pos_label="true")
f1 = f1_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)

print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Precision: 0.0
Recall: 0.0
F1-score: 0.0


In [17]:
pipe = make_pipeline(StandardScaler(), LogisticRegression())

In [18]:
pipe.fit(X_train, Y_train)
pipe_prediction = pipe.predict(X_test)
pipeProbability = pipe.predict_proba(X_test)
pipe_defectiveProbability = pipeProbability[:, 1]

print(pipe_prediction[:20])

print("\nClasses:")
print(pipe.classes_)

print("\nProbability shape:")
print(pipeProbability.shape)

print("\nFirst 5 probability rows:")
print(pipeProbability[:5])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(222, 2)

First 5 probability rows:
[[0.97725754 0.02274246]
 [0.97589319 0.02410681]
 [0.97895044 0.02104956]
 [0.9662071  0.0337929 ]
 [0.94999157 0.05000843]]


In [19]:
pipeAccuracyScore = accuracy_score(Y_test, pipe_prediction)
pipeConfusionMatrix = confusion_matrix(Y_test, pipe_prediction)
pipePrecision = precision_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeRecall = recall_score(Y_test, pipe_prediction, pos_label="true")
pipeF1 = f1_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeROC_AUC = roc_auc_score(Y_test, pipe_defectiveProbability)

print("Accuracy: ", pipeAccuracyScore)
print("\nConfusion Matrix: ")
print(pipeConfusionMatrix)

print("\nPrecision:", pipePrecision)
print("\nRecall:", pipeRecall)
print("\nF1-score:", pipeF1)
print("\nROC-AUC Score:", pipeROC_AUC)

Accuracy:  0.9279279279279279

Confusion Matrix: 
[[205   2]
 [ 14   1]]

Precision: 0.3333333333333333

Recall: 0.06666666666666667

F1-score: 0.1111111111111111

ROC-AUC Score: 0.864573268921095


## Day 3 Findings

The DummyClassifier baseline achieved about 93.2% accuracy but detected 0 defective modules. Logistic Regression achieved about 92.8% accuracy and detected 1 defective modules, but it still missed 14 of the 15 defective modules in the test set. This shows that Logistic Regression does not outperform the baseline in terms of accuracy, but it does improve defect detection by identifying one defective module that the baseline missed. Therefore, accuracy alone does not give a complete picture of the model's performance, especially because the dataset is imbalanced.

## Day 4

In [20]:
from sklearn.ensemble import RandomForestClassifier

In [21]:
randomForest = RandomForestClassifier(random_state=42)

In [22]:
randomForest.fit(X_train, Y_train)

randomForest_prediction = randomForest.predict(X_test)
randomForest_probability = randomForest.predict_proba(X_test)
randomForest_defectiveProbability =randomForest_probability[:, 1]

print(randomForest_prediction[:20])

print("\nClasses:")
print(randomForest.classes_)

print("\nProbability shape:")
print(randomForest_probability.shape)

print("\nFirst 5 probability rows:")
print(randomForest_probability[:5])


['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'true' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(222, 2)

First 5 probability rows:
[[1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]]


In [23]:
randomForest_confusionMatrix = confusion_matrix(Y_test, randomForest_prediction)
randomForest_accuracy = accuracy_score(Y_test, randomForest_prediction)
randomForest_precision = precision_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_recall = recall_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_F1 = f1_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_ROC_AUC = roc_auc_score(Y_test, randomForest_defectiveProbability)

print("Confusion Matrix: ")
print(randomForest_confusionMatrix)
print("\nAccuracy: ", randomForest_accuracy)
print("\nPrecision:", randomForest_precision)
print("\nRecall:", randomForest_recall)
print("\nF1-score:", randomForest_F1)
print("\nROC-AUC Score:", randomForest_ROC_AUC)

Confusion Matrix: 
[[204   3]
 [ 10   5]]

Accuracy:  0.9414414414414415

Precision: 0.625

Recall: 0.3333333333333333

F1-score: 0.43478260869565216

ROC-AUC Score: 0.8750402576489533


## Day 4 Findings

Random Forest achieved higher accuracy (94.1%) than Logistic Regression (92.8%). Random Forest detected 5 of the 15 defective modules, while Logistic Regression detected only 1 of 15. This means Random Forest achieved a higher recall of 33.3%, compared with 6.7% for Logistic Regression.

Random Forest had higher precision, F1-score, and ROC-AUC than Logistic Regression. Therefore, Random Forest performed better than Logistic Regression on PC1 under this train/test split.

However, Random Forest still missed 10 of the 15 defective modules. This means that although Random Forest performed better than Logistic Regression, it still did not detect a large proportion of the actual defects. This shows accuracy alone is not sufficient for evaluating software defect prediction models.

## Day 5

In [24]:
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_validate
from sklearn.metrics import make_scorer, precision_score, recall_score, f1_score

In [25]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [26]:
metric_names = {
    "accuracy": "accuracy",
    "precision": make_scorer(precision_score, pos_label="true"),
    "recall": make_scorer(recall_score, pos_label="true"),
    "f1": make_scorer(f1_score, pos_label="true"),
    "roc_auc": "roc_auc"
}

In [27]:
baseline_cv = cross_validate(estimator=dummy, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(baseline_cv, "\n")

{'fit_time': array([0.0010798 , 0.00050783, 0.00049305, 0.00055075, 0.00041795]), 'score_time': array([0.00733113, 0.00517297, 0.00571084, 0.00474119, 0.00457191]), 'test_accuracy': array([0.92696629, 0.92696629, 0.93220339, 0.93220339, 0.93220339]), 'test_precision': array([0., 0., 0., 0., 0.]), 'test_recall': array([0., 0., 0., 0., 0.]), 'test_f1': array([0., 0., 0., 0., 0.]), 'test_roc_auc': array([0.5, 0.5, 0.5, 0.5, 0.5])} 



/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter

In [28]:
print("Accuracy mean:", baseline_cv["test_accuracy"].mean())
print("Accuracy std:", baseline_cv["test_accuracy"].std())

print("\nPrecision mean:", baseline_cv["test_precision"].mean())
print("Precision std:", baseline_cv["test_precision"].std())

print("\nRecall mean:", baseline_cv["test_recall"].mean())
print("Recall std:", baseline_cv["test_recall"].std())

print("\nF1 mean:", baseline_cv["test_f1"].mean())
print("F1 std:", baseline_cv["test_f1"].std())

print("\nROC-AUC mean:", baseline_cv["test_roc_auc"].mean())
print("ROC-AUC std:", baseline_cv["test_roc_auc"].std())

Accuracy mean: 0.9301085507522376
Accuracy std: 0.0025656434175028112

Precision mean: 0.0
Precision std: 0.0

Recall mean: 0.0
Recall std: 0.0

F1 mean: 0.0
F1 std: 0.0

ROC-AUC mean: 0.5
ROC-AUC std: 0.0


In [29]:
logistic_cv = cross_validate(estimator=pipe, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(logistic_cv, "\n")

{'fit_time': array([0.00524402, 0.00395513, 0.00429797, 0.00416303, 0.00371623]), 'score_time': array([0.0066309 , 0.00620103, 0.00622892, 0.00736213, 0.00622797]), 'test_accuracy': array([0.91011236, 0.92696629, 0.93220339, 0.92655367, 0.93220339]), 'test_precision': array([0.        , 0.5       , 0.        , 0.42857143, 0.        ]), 'test_recall': array([0.        , 0.30769231, 0.        , 0.25      , 0.        ]), 'test_f1': array([0.        , 0.38095238, 0.        , 0.31578947, 0.        ]), 'test_roc_auc': array([0.85128205, 0.83356643, 0.81691919, 0.87272727, 0.83131313])} 



/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


In [30]:
print("Accuracy mean:", logistic_cv["test_accuracy"].mean())
print("Accuracy std:", logistic_cv["test_accuracy"].std())

print("\nPrecision mean:", logistic_cv["test_precision"].mean())
print("Precision std:", logistic_cv["test_precision"].std())

print("\nRecall mean:", logistic_cv["test_recall"].mean())
print("Recall std:", logistic_cv["test_recall"].std())

print("\nF1 mean:", logistic_cv["test_f1"].mean())
print("F1 std:", logistic_cv["test_f1"].std())

print("\nROC-AUC mean:", logistic_cv["test_roc_auc"].mean())
print("ROC-AUC std:", logistic_cv["test_roc_auc"].std())

Accuracy mean: 0.9256078207325589
Accuracy std: 0.008122222115453995

Precision mean: 0.18571428571428572
Precision std: 0.22857142857142856

Recall mean: 0.11153846153846154
Recall std: 0.1378190220551455

F1 mean: 0.13934837092731828
F1 std: 0.1719057101102527

ROC-AUC mean: 0.8411616161616161
ROC-AUC std: 0.01918952040600592


In [31]:
randomForest_cv = cross_validate(estimator=randomForest, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(randomForest_cv, "\n")

{'fit_time': array([0.08626318, 0.08210206, 0.08313394, 0.08583999, 0.09030199]), 'score_time': array([0.01241684, 0.01253176, 0.01229906, 0.01306105, 0.01373482]), 'test_accuracy': array([0.93258427, 0.9494382 , 0.93220339, 0.9039548 , 0.93785311]), 'test_precision': array([0.57142857, 0.83333333, 0.5       , 0.14285714, 0.6       ]), 'test_recall': array([0.30769231, 0.38461538, 0.16666667, 0.08333333, 0.25      ]), 'test_f1': array([0.4       , 0.52631579, 0.25      , 0.10526316, 0.35294118]), 'test_roc_auc': array([0.83496503, 0.88228438, 0.82651515, 0.85429293, 0.81464646])} 



In [32]:
print("Accuracy mean:", randomForest_cv["test_accuracy"].mean())
print("Accuracy std:", randomForest_cv["test_accuracy"].std())

print("\nPrecision mean:", randomForest_cv["test_precision"].mean())
print("Precision std:", randomForest_cv["test_precision"].std())

print("\nRecall mean:", randomForest_cv["test_recall"].mean())
print("Recall std:", randomForest_cv["test_recall"].std())

print("\nF1 mean:", randomForest_cv["test_f1"].mean())
print("F1 std:", randomForest_cv["test_f1"].std())

print("\nROC-AUC mean:", randomForest_cv["test_roc_auc"].mean())
print("ROC-AUC std:", randomForest_cv["test_roc_auc"].std())

Accuracy mean: 0.9312067542690281
Accuracy std: 0.014980783851303704

Precision mean: 0.5295238095238095
Precision std: 0.22334094818980688

Recall mean: 0.23846153846153845
Recall std: 0.10536246569899671

F1 mean: 0.3269040247678019
F1 std: 0.14200345746550302

ROC-AUC mean: 0.8425407925407924
ROC-AUC std: 0.023706422951400057


In [33]:
models = ["DummyClassifier", "Logistic Regression", "Random Forest Classifier"]
meanTableAccuracy = [baseline_cv["test_accuracy"].mean(), logistic_cv["test_accuracy"].mean(), randomForest_cv["test_accuracy"].mean()]
meanTablePrecision = [baseline_cv["test_precision"].mean(), logistic_cv["test_precision"].mean(), randomForest_cv["test_precision"].mean()]
meanTableRecall = [baseline_cv["test_recall"].mean(), logistic_cv["test_recall"].mean(), randomForest_cv["test_recall"].mean()]
meanTableF1 = [baseline_cv["test_f1"].mean(), logistic_cv["test_f1"].mean(), randomForest_cv["test_f1"].mean()]
meanTableRocAuc = [baseline_cv["test_roc_auc"].mean(), logistic_cv["test_roc_auc"].mean(), randomForest_cv["test_roc_auc"].mean()]

meanComparisonTable = pd.DataFrame({
    "Model": models,
    "Accuracy Mean": meanTableAccuracy,
    "Precision Mean": meanTablePrecision,
    "Recall Mean": meanTableRecall,
    "F1 Mean": meanTableF1,
    "ROC-AUC Mean": meanTableRocAuc
})

print(meanComparisonTable)

                      Model  Accuracy Mean  Precision Mean  Recall Mean  \
0           DummyClassifier       0.930109        0.000000     0.000000   
1       Logistic Regression       0.925608        0.185714     0.111538   
2  Random Forest Classifier       0.931207        0.529524     0.238462   

    F1 Mean  ROC-AUC Mean  
0  0.000000      0.500000  
1  0.139348      0.841162  
2  0.326904      0.842541  


In [34]:
stdTableAccuracy = [baseline_cv["test_accuracy"].std(), logistic_cv["test_accuracy"].std(), randomForest_cv["test_accuracy"].std()]
stdTablePrecision = [baseline_cv["test_precision"].std(), logistic_cv["test_precision"].std(), randomForest_cv["test_precision"].std()]
stdTableRecall = [baseline_cv["test_recall"].std(), logistic_cv["test_recall"].std(), randomForest_cv["test_recall"].std()]
stdTableF1 = [baseline_cv["test_f1"].std(), logistic_cv["test_f1"].std(), randomForest_cv["test_f1"].std()]
stdTableRocAuc = [baseline_cv["test_roc_auc"].std(), logistic_cv["test_roc_auc"].std(), randomForest_cv["test_roc_auc"].std()]

stdComparisonTable = pd.DataFrame({
    "Model STD": models,
    "Accuracy STD": stdTableAccuracy,
    "Precision STD": stdTablePrecision,
    "Recall STD": stdTableRecall,
    "F1 STD": stdTableF1,
    "ROC-AUC STD": stdTableRocAuc
})

print(stdComparisonTable)

                  Model STD  Accuracy STD  Precision STD  Recall STD  \
0           DummyClassifier      0.002566       0.000000    0.000000   
1       Logistic Regression      0.008122       0.228571    0.137819   
2  Random Forest Classifier      0.014981       0.223341    0.105362   

     F1 STD  ROC-AUC STD  
0  0.000000     0.000000  
1  0.171906     0.019190  
2  0.142003     0.023706  


In [35]:
finalComparisonTable = pd.DataFrame({
    "Model": models,
    "Accuracy": [f"{meanTableAccuracy[i]:.3f} ± {stdTableAccuracy[i]:.3f}" for i in range(3)],
    "Precision": [f"{meanTablePrecision[i]:.3f} ± {stdTablePrecision[i]:.3f}" for i in range(3)],
    "Recall": [f"{meanTableRecall[i]:.3f} ± {stdTableRecall[i]:.3f}" for i in range(3) ],
    "F1": [f"{meanTableF1[i]:.3f} ± {stdTableF1[i]:.3f}" for i in range(3)],
    "ROC-AUC": [f"{meanTableRocAuc[i]:.3f} ± {stdTableRocAuc[i]:.3f}" for i in range(3)]
})

print(finalComparisonTable)

                      Model       Accuracy      Precision         Recall  \
0           DummyClassifier  0.930 ± 0.003  0.000 ± 0.000  0.000 ± 0.000   
1       Logistic Regression  0.926 ± 0.008  0.186 ± 0.229  0.112 ± 0.138   
2  Random Forest Classifier  0.931 ± 0.015  0.530 ± 0.223  0.238 ± 0.105   

              F1        ROC-AUC  
0  0.000 ± 0.000  0.500 ± 0.000  
1  0.139 ± 0.172  0.841 ± 0.019  
2  0.327 ± 0.142  0.843 ± 0.024  


## Day 5 Findings

The DummyClassifier achieved about 93.0% accuracy, but it had zero precision, recall, and F1 with a ROC-AUC of 0.5. This shows that the high accuracy comes from predicting the majority class while failing to detect defective modules. Both Logistic Regression and Random Forest performed better than the DummyClassifier for defect detection.
Random Forest had higher precision (0.530), recall (0.238), F1-score (0.327), and slightly higher ROC-AUC (0.843) than Logistic Regression. Therefore, Random Forest was stronger overall for defect detection under the 5-fold cross-validation evaluation. However, its mean recall of about 24% shows that it still missed many defective modules. This again shows why accuracy alone is not sufficient for evaluating software defect prediction models on an imbalanced dataset.


## Day 6

In [36]:
randomForest_class_weight = RandomForestClassifier(random_state=42, class_weight="balanced")

In [37]:
randomForest_class_weight.fit(X_train, Y_train)

randomForest_prediction_class_weight = randomForest_class_weight.predict(X_test)
randomForest_probability_class_weight = randomForest_class_weight.predict_proba(X_test)
randomForest_defectiveProbability_class_weight =randomForest_probability_class_weight[:, 1]

print(randomForest_prediction_class_weight[:20])

print("\nClasses:")
print(randomForest_class_weight.classes_)

print("\nProbability shape:")
print(randomForest_probability_class_weight.shape)

print("\nFirst 5 probability rows:")
print(randomForest_probability_class_weight[:5])


['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'true' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(222, 2)

First 5 probability rows:
[[1.   0.  ]
 [0.99 0.01]
 [1.   0.  ]
 [0.98 0.02]
 [1.   0.  ]]


In [39]:
randomForest_confusionMatrix_class_weight = confusion_matrix(Y_test, randomForest_prediction_class_weight)
randomForest_accuracy_class_weight = accuracy_score(Y_test, randomForest_prediction_class_weight)
randomForest_precision_class_weight = precision_score(Y_test, randomForest_prediction_class_weight, pos_label="true", zero_division=0)
randomForest_recall_class_weight = recall_score(Y_test, randomForest_prediction_class_weight, pos_label="true", zero_division=0)
randomForest_F1_class_weight = f1_score(Y_test, randomForest_prediction_class_weight, pos_label="true", zero_division=0)
randomForest_ROC_AUC_class_weight = roc_auc_score(Y_test, randomForest_defectiveProbability_class_weight)

print("Confusion Matrix: ")
print(randomForest_confusionMatrix_class_weight)
print("\nAccuracy: ", randomForest_accuracy_class_weight)
print("\nPrecision:", randomForest_precision_class_weight)
print("\nRecall:", randomForest_recall_class_weight)
print("\nF1-score:", randomForest_F1_class_weight)
print("\nROC-AUC Score:", randomForest_ROC_AUC_class_weight)

Confusion Matrix: 
[[202   5]
 [  7   8]]

Accuracy:  0.9459459459459459

Precision: 0.6153846153846154

Recall: 0.5333333333333333

F1-score: 0.5714285714285714

ROC-AUC Score: 0.8964573268921096


## Day 6 Findings:

The balanced Random Forest detected 8 of the 15 defective modules, compared with 5 of 15 for the original Random Forest. Class weighting increased recall from 0.333 to 0.533 and reduced false negatives from 10 to 7, showing that the model became better at detecting defective modules. However, precision decreased slightly from 0.625 to 0.615 because false positives increased from 3 to 5. The F1-score improved from 0.435 to 0.571, while accuracy increased slightly from 0.941 to 0.946 and ROC-AUC increased from 0.875 to 0.896. Therefore, class weighting improved defect detection by increasing recall and F1-score, while also slightly improving accuracy and ROC-AUC.